### ISSUE TITLE
Z-Image text sequence length issue

### ISSUE THREAD

**Issue by dxqb** [huggingface/diffusers#12893]
### Describe the bug

I think there might be an issue with calculating the sequence length of `cap_feat` (which is the text encoder output), and masking it accordingly.

I'm going to use code links from *before* the omni commit, because it's easier to read - but the issue seems exist in both, before and after the omni commit.

https://github.com/huggingface/diffusers/blob/52766e6a6939ac6e74375bde5e19c5e0b90d24c1/src/diffusers/models/transformers/transformer_z_image.py#L593-L611

In these lines, the sequence length for each `cap_feats` sample is put into `cap_item_seqlens` and is then used to create an attention mask in line 611.

The problem with this is that `cap_feats` has been overwritten before in these lines:
https://github.com/huggingface/diffusers/blob/52766e6a6939ac6e74375bde5e19c5e0b90d24c1/src/diffusers/models/transformers/transformer_z_image.py#L552-L560

`cap_feats` therefore isn't anymore what the caller has passed. instead, `cap_feats` has already been padded by `patchify_and_embed`. Therefore, *all* `cap_item_seqlens` are identical to `cap_max_item_seqlen`.

This leads to unmasked text tokens during attention, which probably wasn't the intention here.

### Reproduction

put a breakpoint in this line:
https://github.com/huggingface/diffusers/blob/52766e6a6939ac6e74375bde5e19c5e0b90d24c1/src/diffusers/models/transformers/transformer_z_image.py#L612

all `cap_items_seqlens` are identical, `cap_attn_mask` is all True - even if multiple different text lengths were passed to the transformer.

### Logs

```shell

```

### System Info

-

### Who can help?

@JerryWu-code @RuoyiDu @yiyixuxu

**Comment by dxqb** [comment]
this should still be adressed @JerryWu-code @RuoyiDu

**Comment by Mr-Neutr0n** [comment]
Opened a fix for this in #13131.

The root cause is that `_prepare_sequence` computes `item_seqlens` from the already-padded features (padded to multiples of `SEQ_MULTI_OF` by `_pad_with_ids`), so the attention mask treats all inner padding positions as valid. The fix uses the existing `inner_pad_mask` to exclude those positions from the attention mask, both in the individual refiner stages and in the unified sequence passed to the main transformer layers.

**Comment by ChrisLiu6** [comment]
Hi guys, could you kindly take a look at this issue https://github.com/Tongyi-MAI/Z-Image/issues/131 and check if this solves your questions?

**Comment by dxqb** [comment]
> Hi guys, could you kindly take a look at this issue [Tongyi-MAI/Z-Image#131](https://github.com/Tongyi-MAI/Z-Image/issues/131) and check if this solves your questions?

Thanks, but no. I think this is a bug in the `diffusers` z-image code. It may or may not exist in your original inference code, but it does in `diffusers`.

For reproduction, to be sure, let's put a breakpoint right at attention. For example here:
https://github.com/huggingface/diffusers/blob/aac94befce95f1842511cf13f43b2b60fa27b966/src/diffusers/models/transformers/transformer_z_image.py#L133


Run this code, and note the two prompts of different lengths:
```
import torch
from diffusers import ZImagePipeline
from diffusers.utils import load_image

pipe = ZImagePipeline.from_pretrained("Tongyi-MAI/Z-Image-Turbo", torch_dtype=torch.bfloat16)
pipe.to("cuda")

image = pipe(
    ["a very very very very very very very very very very very very very very long prompt", "short prompt"],
    width=512,
    height=512,
    num_inference_steps=9,
    guidance_scale=0.0,
    generator=torch.Generator("cuda").manual_seed(42),
).images
image[0].save("zimage-0.png")
image[1].save("zimage-1.png")
```

Then, `cont` a couple of times until you're past the refiner layers, in the regular layers that use joint attention.
```
(Pdb) hidden_states.shape
torch.Size([2, 1056, 3840])
```
batch size 2, 1024 image tokens, apparently 32 text tokens.

```
(Pdb) attention_mask.shape
torch.Size([2, 1, 1, 1056])
```

matches

```
(Pdb) attention_mask.sum()
tensor(2112, device='cuda:0')
(Pdb) torch.any(attention_mask == False)
tensor(False, device='cuda:0')
```

here is the problem. not a single token is masked.
quoting your link:
> On the other hand, we have also introduced "outer padding", which is to pad samples in a batch to the same length. This kind of padding does not contribute to attention computation

padding does contribute to attention, because the mask is a no-op

**Comment by dxqb** [comment]
please ignore my previous (now deleted) message. this was before I understood your answer.

> Hi guys, could you kindly take a look at this issue [Tongyi-MAI/Z-Image#131](https://github.com/Tongyi-MAI/Z-Image/issues/131) and check if this solves your questions?

Thank you. That's a smart idea.
I guess to avoid masking, but still be able to have variable caption lengths in the same batch during training. And without having a mismatch between training and inference by training the model to deal with the "inner padding".

The only thing I'd suggest then is to actually use the benefit of this. The current code always uses the attention mask in attention, even if it's all-True and therefore a no-op. This causes torch SDPA to not use its internal flash attention algorithm, but the slower one that supports masks.

I have prepared a small PR here: https://github.com/huggingface/diffusers/pull/13184
Please let me know if you agree.

It's about 15% faster for inference(even for batch size 1). Usually more in training but I haven't benchmarked it now.
The inference result is pixel-identical.

**Comment by ChrisLiu6** [comment]
Thanks! The PR looks good to me.

**Comment by Beinsezii** [comment]
Probably also fixed by #12955

**Comment by yiyixuxu** [comment]
should be fixed now we merged https://github.com/huggingface/diffusers/pull/12955
